### PR TITLE
Prevent components from working with non-uploaded files

### PR DESCRIPTION
**Pull request by aliabid94**

### CODE REVIEW

**Review by freddyaboulton — COMMENTED** [review]
I think we need to modify `run_fn_batch` to accept explicit_call (https://github.com/gradio-app/gradio/blob/cc51d9063043d896dbd219787c32e601f2fadd3e/gradio/blocks.py#L1532) and also this call to preprocess in the non batch case - https://github.com/gradio-app/gradio/blob/cc51d9063043d896dbd219787c32e601f2fadd3e/gradio/blocks.py#L1615

Otherwise looks good to me!

File: gradio/blocks.py [hunk 1]
```diff
@@ -1298,7 +1299,11 @@ def validate_inputs(self, fn_index: int, inputs: list[Any]):
             )
 
     def preprocess_data(
-        self, fn_index: int, inputs: list[Any], state: SessionState | None
+        self,
+        fn_index: int,
+        inputs: list[Any],
+        state: SessionState | None,
+        explicit_call: bool = False,
```
Comment (akx) [inline gradio/blocks.py:1306]: New argument not documented?

**Review by akx — CHANGES_REQUESTED** [review]

File: gradio/processing_utils.py [hunk 2]
```diff
@@ -264,6 +267,9 @@ def _move_to_cache(d: dict):
             payload.path = payload.url
         elif not block.proxy_url:
             # If the file is on a remote server, do not move it to cache.
+            if check_in_upload_folder:
+                path = os.path.abspath(payload.path)
+                assert path.startswith(get_upload_folder()), "File not in upload folder"
```
[the remark this answers is not in the document]
Comment (abidlabs) [reply gradio/processing_utils.py:272]: Use this function instead, its more robust:

https://vscode.dev/github/gradio-app/gradio/blob/revert/gradio/utils.py#L943

File: gradio/processing_utils.py [hunk 3]
```diff
@@ -252,6 +253,8 @@ def move_files_to_cache(
         block: The component whose data is being processed
         postprocess: Whether its running from postprocessing
         root_url: The root URL of the local server, if applicable
+        add_urls: Whether to add URLs to the payload
+        check_in_upload_folder: If True, instead of moving the file to cache, checks if the file is in upload folder.
```
[the remark this answers is not in the document]
Comment (abidlabs) [reply gradio/processing_utils.py:257]: Yes agree with @akx

File: gradio/utils.py [hunk 4]
```diff
@@ -1082,3 +1083,9 @@ def compare_objects(obj1, obj2, path=None):
         return edits
 
     return compare_objects(old, new)
+
+
+def get_upload_folder():
```
[the remark this answers is not in the document]
Comment (abidlabs) [reply gradio/utils.py:1088]: Now that we have this function, we no longer need this attribute on the `Block` class. We can just use `utils.get_upload_folder()` everywhere this is used:

https://github.com/gradio-app/gradio/blob/99af290a9683f65e449b06a5e07572583ebcaaed/gradio/blocks.py#L122

File: .changeset/moody-impalas-rule.md [hunk 5]
```diff
@@ -0,0 +1,5 @@
+---
+"gradio": minor
```
Comment (abidlabs) [inline .changeset/moody-impalas-rule.md:2]: Should be a patch version

File: gradio/processing_utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (abidlabs) [reply gradio/processing_utils.py:272]: And maybe we can just `return` after this line (makes the logic easier to follow imo)

**Comment by abidlabs** [comment]
Thanks @aliabid94! This should fix the issue. I would just add a unit test as well, and there's a few nits above.

**Review by abidlabs — APPROVED** [review]

File: gradio/blocks.py [hunk 6]
```diff
@@ -1109,6 +1109,7 @@ def __call__(self, *inputs, fn_index: int = 0, api_name: str | None = None):
             inputs=processed_inputs,
             request=None,
             state={},
+            explicit_call=True,
```
Comment (abidlabs) [inline gradio/blocks.py:1112]: Nice catch

**Comment by PinkDraconian** [comment]
This fix looks great. Thank you for the resolution. 
Would you also be able to "accept" my report? That way the process over there moves along.

File: gradio/processing_utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (aliabid94) [reply gradio/processing_utils.py:272]: > maybe we can just return after this line

I thought that the add_urls logic might still be needed, so I don't return early

[the remark this answers is not in the document]
Comment (abidlabs) [reply gradio/processing_utils.py:272]: Oh you're right, I missed that.

**Pull request merged**